tests: introduce a proper JSON diff for topotests

GalaxyGorilla opened the pull request:
Background: I'm currently working on OSPF SR topotests and
handling large'ish JSON files which has become quite a pain.

Diff'ing JSON objects is a crucial operation in the topotests for
comparing e.g. vtysh output (formatted as JSON) with a file which
covers the expectation of the tests. The current diff functionality
is 'self-written' and intended to test a JSON object d2 on being a
subset of another JSON object d1. For mismatches a diff is generated
based on a normalized textual representation of the JSON objects.

This approach has several disadvantages:

  * the human provided JSON text might not be normalized, hence
    a diff with line numbers might be worthless since it provides
    close to zero orientation what the problem is
  * the diff contains changes like commatas which are meaningless
  * the diff might contain a lot of changes about meaningless
    content which is present in d1 but not in d2
  * there is no proper functionality to test for 'equality' of
    d1 and d2
  * it is not possible to test for order, e.g. JSON arrays are
    just tested with respect to being a subset of another array
  * it is not possible to check if a key exists without also
    checking the value of that particular key

This commit attempts to solve these issues. An error report is
generated which includes the "JSON Path" to the problematic JSON
elements and also hints on what the actual problem is (e.g. missing
key, mismatch in dict values etc.).

A special parameter 'exact' was introduced such that equality can be
tested. Also there was a convention that absence of keys can be
tested using the key in question with value 'None'. This convention
is still honored such that full backwards compatiiblity is in
place.

Further order can be tested using the new tag '__ordered__' in
lists (as first element). Example:

    d1 = [1, 2, 3]
    d2 = ['__ordered__', 1, 3, 2]

Tesing d1 and d2 this way will now result in an error.

Key existence can now be tested using an asterisk '*'. Example:

    d1 = [1, 2, 3]
    d2 = [1, '*', 3]

    d1 = {'a': 1, 'b': 2}
    d2 = {'a': '*'}

Both cases will result now in a clean diff for d1 and d2.

Signed-off-by: GalaxyGorilla <sascha@netdef.org>

GalaxyGorilla commented:
Currently set to WIP since there are some bugs left. To run a topotest one needs to install `jsondiff` via `pip install jsondiff`. I'm not sure how much of a No-Go a new dependency is. Also I'd like to know if there are Python related conventions (e.g. PEP 8) that I have to care for.

Here is an example output from the new JSON diff error report:

```
E       AssertionError: "rt1" JSON output mismatches the expected result
E       assert Generated JSON diff error report:
E
E         > $->10.0.1.0/24[0]->nexthops[1]: Key 'afi' missing in d1
E         > $->10.0.1.0/24[0]->nexthops: List element with index '0' missing in d1. This index has value:
E
E         	{
E         	    "ip": "10.0.1.2",
E         	    "afi": "ipv4",
E         	    "interfaceName": "eth-sw1"
E         	}
E
E         > $->10.0.2.0/24[0]->distance: Expected '110' in d1 but got '115' in d2
```

rzalamena commented:
@GalaxyGorilla nice improvement, the new output looks much better.

One way to test if your new implementation is working 100% is to run these tests:
https://github.com/FRRouting/frr/blob/master/tests/topotests/lib/test/test_json.py

I wrote it to test the implementation and in case we ever change how it works (this JSON diffing is not conventional as we allow subsets like you mentioned).

Also, if more people find this useful, I think you should make the dependency optional. If the user has the package then use it otherwise fallback to old implementation (this way we can get the PR merged faster).

GalaxyGorilla commented:
@rzalamena Thanks for the feedback! Since I just got some unexpected errors about list handling with the new diff I had a look into those tests and found this here:

https://github.com/FRRouting/frr/blob/master/tests/topotests/lib/test/test_json.py#L186

Does that mean that _also_ lists are just tested for being a subset of another list?

rzalamena commented:
@GalaxyGorilla the tests `test_json_with_list_sucess` try the following:

1. test if object is of list type: anything else will fail
2. test if list item matches what we expected (something like obj == subset of obj)
3. test list size: we might match empty list, half empty list or full list, however we can't match if list have more items than expected.

Right after these tests we also have `test_json_with_list_failure`, which try the following:

1. test if object of non-list type causes error
2. test if comparison fails when list object is different
3. test if list list size comparison fails when we have more items than expected

> Does that mean that also lists are just tested for being a subset of another list? Does order matter in any way?

Yes, we accept subsets in lists. No, order does not matter.

GalaxyGorilla commented:
@rzalamena But how do you test _anything_ that relies on order like e.g. label stacks (an ordered list of MPLS labels) in segment routing?

rzalamena commented:
@GalaxyGorilla 
> But how do you test _anything_ that relies on order like e.g. label stacks (an ordered list of MPLS labels) in segment routing?

it wasn't a requirement until now. How about a new diff flag to signalize order? Maybe a second function to test list order? Testing order by default might be a bad idea at this point and this looks like a niche test.

GalaxyGorilla commented:
> @GalaxyGorilla
> 
> > But how do you test _anything_ that relies on order like e.g. label stacks (an ordered list of MPLS labels) in segment routing?
> 
> it wasn't a requirement until now. How about a new diff flag to signalize order? Maybe a second function to test list order? Testing order by default might be a bad idea at this point and this looks like a niche test.

Ugh ... so I see two paths which we can go here in general:

1. Throw away this PR and create a new one which builds on the current JSON diff code with better visual diffing (like in this PR) and features an additional flag for caring about order.
2. Adjust all JSON related APIs to use dicts where order doesn't matter and lists where order matters.

Not sure which way to go. I can imagine that people are using the `show ... json` commands in their scripts. So option 1 appears to be the way to go.

Maybe I can also implement this somehow using the current code, I need to check ...

rzalamena commented:
@GalaxyGorilla 

> 1. Throw away this PR and create a new one which builds on the current JSON diff code with better visual diffing (like in this PR) and features an additional flag for caring about order.

You don't need to throw away the PR (it contains some useful information for reviewers later - e.g. this discussion), just mark it as draft (I can't I'm not the owner) and maybe the WIP label.

Change the code in the branch, rebase or ammend the commit and force push your branch.

> 2. Adjust all JSON related APIs to use dicts where order doesn't matter and lists where order matters.

This is not realistic... And I also think most of the time lists order shouldn't matter.

---

About the function change: you don't need to change all tests function call, just add an optional parameter to the compare function (e.g. `lists_ordered=False`).

GalaxyGorilla commented:
@rzalamena I thought a bit about our problem here and I think it is a bit more complicated than just adding another parameter to the compare function. Here is an example from `show ip route`:

```
  "3.3.3.3\/32":[
    {
      "prefix":"3.3.3.3\/32",
      "protocol":"ospf",
      "selected":true,
      "destSelected":true,
      "distance":115,
      "metric":20,
      "installed":true,
      "nexthops":[
        {
          "fib":true,
          "ip":"10.0.1.3",
          "afi":"ipv4",
          "interfaceName":"eth-sw1",
          "active":true,
          "labels":[
            17030,
            17040
          ]
        }
      ]
    }
  ]
```

As you can see the `nexthops` key has an (unordered) list as value. And this list again contains the 'labels' key with an ordered list. Hence for testing it wouldn't suffice to just completely test with `lists_ordered=True` as  parameter to the compare function because that might break the comparison on the `nexthops` list. How about a key like `"__ordered__"=true` which you put into the list which should be tested for order? 

This is really getting an ugly problem ... with all these special cases I'm not sure anymore if its even worth to drag in this `jsondiff` dependency.

qlyoung commented:
@GalaxyGorilla I have an project unrelated to FRR that needs JSON diffs as well. Here's what I use there:

```python
from difflib import unified_diff

a = <some json output>
b = <some json output>

json_a = json.dumps(json.loads(a), sort_keys=True, indent=4)
json_b = json.dumps(json.loads(b), sort_keys=True, indent=4)

print(unified_diff(json_a.split("\n"), json_b.split("\n"), lineterm="", fromfile="jsonA", tofile="jsonB"))
```

It works quite well and produces nice unified diffs. Maybe that would work here?

GalaxyGorilla commented:
> @GalaxyGorilla I have an project unrelated to FRR that needs JSON diffs as well. Here's what I use there:
> 
> ```python
> from difflib import unified_diff
> 
> a = <some json output>
> b = <some json output>
> 
> json_a = json.dumps(json.loads(a), sort_keys=True, indent=4)
> json_b = json.dumps(json.loads(b), sort_keys=True, indent=4)
> 
> print(unified_diff(json_a.split("\n"), json_b.split("\n"), lineterm="", fromfile="jsonA", tofile="jsonB"))
> ```
> 
> It works quite well and produces nice unified diffs. Maybe that would work here?

That looks certainly interesting, but what we need here is something that copes well with subsets and optional order in arrays. Also there is this thing with checking presence of keys. I agreed with @rzalamena (in slack) that I will give the whole JSON diff code an overhaul without adding another dependency to get something tailored to our needs.

GalaxyGorilla commented:
@rzalamena I updated the PR description and the tests. Can you check the tests to see if things like the new `'__ordered__'` tag or the asterisk `'*'` placeholder are "ok"? I'm especially critical about the `'__ordered__'` tag but I see no other way :(

rzalamena commented on tests/topotests/lib/topotest.py:
```diff
@@ -66,144 +67,140 @@ def has_errors(self):
-        result.add_error(
-            "{} value is different (\n{})".format(parent, json_diff(list1, list2))
-        )
+def gen_json_diff_report(d1, d2, exact=False, path="> $", acc=(0 , "")):
```
Would mind adding a small description to this function? Maybe tell the reader this is an internal function.

Other than that I think this is mostly it, the new formatter might want to change some long lines here.

GalaxyGorilla replied:
Done.

I used the python string formatter whereever it makes sense and re-formatted the file using `black`.

rzalamena commented on tests/topotests/lib/topotest.py:
```diff
@@ -66,144 +67,140 @@ def has_errors(self):
-def json_cmp(d1, d2):
+
+def json_cmp(d1, d2, exact=False):
     """
```
Can you document `exact` and tell the reader about `'*'` and `'__ordered__'`?

I think the choices you've made for `'*'` and `'__ordered__'` are good. For an example JSON users unofficially recommend `'__comment__'` for commentaries since the language itself does not provide.

GalaxyGorilla replied:
Done.

rzalamena commented on tests/topotests/lib/topotest.py:
```diff
@@ -218,12 +215,12 @@ def router_output_cmp(router, cmd, expected):
 
-def router_json_cmp(router, cmd, data):
+def router_json_cmp(router, cmd, data, exact=False):
     """
```
Might be useful to document `exact` here as well (or point to `json_cmp`).

GalaxyGorilla replied:
I added some more docu there.

GalaxyGorilla commented:
Here is one more example output:

```
E       AssertionError: "rt1" JSON output mismatches the expected result
E       assert Generated JSON diff error report:
E
E         > $->10.0.1.0/24: d2 has the following element at index 0 which is not present in d1:
E
E           {
E               "prefix": "10.0.1.0/24",
E               "protocol": "ospf",
E               "nexthops": [
E                   {
E                       "ip": "10.0.1.2",
E                       "afi": "ipv4",
E                       "interfaceName": "eth-sw1"
E                   },
E                   {
E                       "ip": "10.0.1.3",
E                       "afi": "ipv4",
E                       "interfaceName": "eth-sw1"
E                   }
E               ]
E           }
E
E           Closest match in d1 is at index 0 with the following errors:
E
E           > $->10.0.1.0/24[0]->nexthops: d1 has 'Array' of length 1 but in d2 it is of length 2
E
E         > $->6.6.6.6/32: d2 has the following element at index 0 which is not present in d1:
E
E           {
E               "destSelected": true,
E               "protocol": "ospf",
E               "selected": true,
E               "installed": true,
E               "prefix": "6.6.6.6/32",
E               "nexthops": [
E                   {
E                       "afi": "ipv4",
E                       "ip": "10.0.1.2",
E                       "labels": [
E                           16060
E                       ],
E                       "fib": true,
E                       "active": true,
E                       "interfaceName": "eth-sw1"
E                   },
E                   {
E                       "afi": "ipv4",
E                       "ip": "10.0.1.3",
E                       "labels": [
E                           17060
E                       ],
E                       "fib": true,
E                       "active": true,
E                       "interfaceName": "eth-sw1"
E                   }
E               ]
E           }
E
E           Closest match in d1 is at index 0 with the following errors:
E
E           > $->6.6.6.6/32[0]->nexthops: d2 has the following element at index 0 which is not present in d1:
E
E               {
E                   "afi": "ipv4",
E                   "ip": "10.0.1.2",
E                   "labels": [
E                       16060
E                   ],
E                   "fib": true,
E                   "active": true,
E                   "interfaceName": "eth-sw1"
E               }
E
E               Closest match in d1 is at index 0 with the following errors:
E
E               > $->6.6.6.6/32[0]->nexthops[0]: d2 has key 'labels' which is not present in d1
E
E           > $->6.6.6.6/32[0]->nexthops: d2 has the following element at index 1 which is not present in d1:
E
E               {
E                   "afi": "ipv4",
E                   "ip": "10.0.1.3",
E                   "labels": [
E                       17060
E                   ],
E                   "fib": true,
E                   "active": true,
E                   "interfaceName": "eth-sw1"
E               }
E
E               Closest match in d1 is at index 1 with the following errors:
E
E               > $->6.6.6.6/32[0]->nexthops[1]: d2 has key 'labels' which is not present in d1
```

As you can see the report talks about a closest match because by default the JSON diff doesn't care about order. I implemented a stupid heuristic which tries to find a 'closest match' to serve a good hint on what went wrong.
